wireguard: add procd init service

feckert opened the pull request:
This pullrequest adds some improvements and a fix for the wireguard proto handling.
I have already send some patches to the mailinglist but no one felt addressed more or less. @jow-.
So I try it with a pullrequest.

Changes already fixed:
* wireguard: skip peer config if public key of the peer is not defined
[Already send via Mailinglist](https://patchwork.ozlabs.org/patch/1204567/)
**[Is already fixed in openwrt master](https://github.com/openwrt/openwrt/commit/7151054abd5d43bf4c623e311a4c1046af4fff94#diff-781e06caaf632d4cf7d4b6eff8c854d2)**
- wireguard: fix interface remove on lonely peers
[Already send via Mailinglist](https://patchwork.ozlabs.org/patch/1204578/)
difference from Mailinglist is that I changed **procd** `boot()` to `start_service()`
**[Is already fixed in LuCI master](https://github.com/openwrt/luci/commit/890dcac2de7b1d6bc6768c69ea7129d27a809ef9#diff-10b0bc939a6ce69b067b6c9c7990c677)**

Changes still pending:
* wireguard: add procd firewall3 inbound rule support
Add an additional uci option **allowed_zones** so we can open the Configured UDP on this firewall zone.
* wireguard: fix reload config on peer change
[Already send via Mailinglist](https://patchwork.ozlabs.org/patch/1202476/)
difference from Mailinglist is that I changed **procd** `boot()` to `start_service()`

feckert committed:
wireguard-tools: fix reload config on peer change

If we change a peer section, then the interface of netifd gets not
reloaded. Because the change were not made in an interface section.
And so the netifd does not recognize the change. And the new config gets
not applied until we do a network restart or we restart the interface
with 'ifup <name>'.

With this new wireguard init script, a md5sum will be calculated on
every network change. The sum is generated over the wireguard peers for
each wireguard interface. If a change in the peers section gets detected
then only the detecated wireguard interface gets restarted.

With this change we can see if the peer section has changed to the
corresponding interface. The wireguard configuration is rewritten and
reconfigured by the netif proto handler.

Signed-off-by: Florian Eckert <fe@dev.tdt.de>

feckert committed:
wireguard-tools: add procd firwall3 inbound rule support

Wireguard does listen on all interfaces. Normally we have to open the
configured UDP port in the firewall in a separate task.
For this to work automatically a firewall rule must be stored in the procd
which opens this for use. So we do not need an additonal firewall configuration
task. Every thinks work automatically on change.

For this to work an additional uci list option ( allowed_zones ) is required
to specify the firewall zone on which the UDP port firewall rule gets applied.

Signed-off-by: Florian Eckert <fe@dev.tdt.de>

ynezz commented:
@zx2c4 can you check this?

zx2c4 commented:
Thanks for sending this my way. I'll look at it next time I'm at a computer, tonight or tomorrow.

feckert committed:
wireguard-tools: update PKG_RELEASE

Signed-off-by: Florian Eckert <fe@dev.tdt.de>

feckert commented:
@zx2c4 I have rebased and update my pullrequest so it should apply correct

blogic commented:
needs another rebase and there is an extra "i" in the first commits title
this is also the missing patch that the one on the ML depends on i guess

feckert commented:
@blogic Thanks for your feedback I have rebased my pullrequest and removed the already applied commit from the mailinglist https://github.com/openwrt/openwrt/commit/7151054abd5d43bf4c623e311a4c1046af4fff94
@jow- mentioned on the mailing list that there should be a better solution. So far I have only noticed this one thing. I have not tried every config combination.

Also see my notes for the commit https://github.com/openwrt/openwrt/pull/2625/commits/6c62662f93c43dd04f6901f7b7f4ebfa8b04f6ed on the mailing list https://lists.infradead.org/pipermail/openwrt-devel/2020-January/021253.html

feckert commented:
@blogic I have update my pullrequest, so this applies cleanly again.
@zx2c4 Is there any chance that this change could go upstream?
I did a second rebase for this pullrequest.

blocktrron commented on package/network/utils/wireguard-tools/files/wireguard.init:
```diff
@@ -42,4 +42,25 @@ start_service() {
+		# delete peer if iface is not present anymore
+		if ! uci -q show "network.${iface}" 1>/dev/null 2>/dev/null; then
+			uci -q delete "network.@${peer}[-1]"
+			uci commit network
```
In case we delete more than one peer, this would commit multiple commits. 

Better move the commit out of this loop and only call it if a peer was deleted.

feckert replied:
I will remove the said section as requested by jow-

blocktrron reviewed (commented):
Changes look sensible to me. Found one minor quirk to improve on.

There are two typos in your commit messages:

` wireguard-tools: add procd firwall3 inbound rule support `
 - firewall3 misspelled
 - `Every thinks work automatically.` --> Everything

jow- commented on package/network/utils/wireguard-tools/files/wireguard.init:
```diff
@@ -0,0 +1,122 @@
+
+		# delete peer if iface is not present anymore
+		if ! uci -q show "network.${iface}" 1>/dev/null 2>/dev/null; then
+			uci -q delete "network.@${peer}[-1]"
```
I still think that is is a no-go to modify (and commit!) network configuration from an init script. This has unpredictable side effects, might accidentally apply intentionally uncommitted changes, destroy formatting and comments in the configuration etc.

feckert replied:
Now that you mention it! I dimly remember you commenting on the mailing list before. But then how am I supposed to do that? The way I see it, it'll have to happen in the LuCI.

feckert replied:
This Is done now in LuCI
openwrt-19.07:
https://github.com/openwrt/luci/commit/86d0bf86eabd04ba397a7e20ea84f892be8ec934

master:
https://github.com/openwrt/luci/commit/890dcac2de7b1d6bc6768c69ea7129d27a809ef9

zx2c4 commented on package/network/utils/wireguard-tools/files/wireguard.init:
```diff
@@ -0,0 +1,122 @@
+
+	. /lib/functions/network.sh
+
+	n_sum="$(md5sum "${cfile}" | cut -d" " -f1)"
```
Why is there md5 here?

feckert replied:
I would have needed this to determine if the configuration has changed, to delete peer section when the inteface is deleted. But as @jow- already recognized this is not ok https://github.com/openwrt/openwrt/pull/2625/files/c0befebe47cd917e1b3a100dc58c9d848744083f#r378338317

zx2c4 replied:
Still doesn't seem like a good reason for md5.

rmilecki replied:
While I'm aware that MD5 isn't good for various security tasks it should still be just fine for simple cases like comparing configs.

If you check `/sbin/reload_config` it's exactly how comparing config files it handled. Using `md5sum`.

zx2c4 commented:
```
If we change a peer section, then the interface of netifd gets not
reloaded. Because the change were not made in an interface section.
And so the netifd does not recognize the change. And the new config gets
not applied until we do a network restart or we restart the interface
with 'ifup <name>'.

With this new wireguard init script, a md5sum will be calculated on
every network change. The sum is generated over the wireguard peers for
each wireguard interface. If a change in the peers section gets detected
then only the detecated wireguard interface gets restarted.

With this change we can see if the peer section has changed to the
corresponding interface. The wireguard configuration is rewritten and
reconfigured by the netif proto handler.
```

Can't you just use `wg syncconf`?

feckert commented:
@zx2c4 The command `wg syncconf`must be new. I haven't heard that one before.

With an interface section, it's no problem. If there is a change in the configuration, the neifd is triggered and reconfigure the wireguard interface. The problem is that we don't notice if something changes in the peer sections. Nobody notices that. That's why I calculate the md5sum to detect a change and make the netifd rewrite the configuration on a ifup command.
With your suggestion `wg syncconf` I already need a new configuration, which the protocol handler wireguard creates. I don't think that helps me here.

zx2c4 commented:
Looks like we're still talking about MD5. That doesn't bode well.

Given that and the other commenters' concerns, I think we should n'ack this patchset.

feckert commented:
I would like to point out that these are a total of 3 separate changes.

1. If the interface is deleted, the peers remain in the configuration. I would have solved this with MD5. This would delete the peers that are no longer needed. In this case the configuration is written which is not OK what the others have noted.

2. When a peer configuration is changed (LuCI), I have to manually restart the interface so that the configuration of the netifd can be applied via its proto handler. Which is not the case with an interface configuration change! I would have solved this also via MD5 to get a change.  Nothing is written to the config. The MD5 is only used to track and apply the changes in the peersections.

3. The other is a feature, like in other services, to not create a firewall rule in a separate step. If a list port is added in the wireguard configuration, this port will be opened automatically.

feckert commented:
Does no one want to look at this again is that your last word?
With this pullrequest two things are added.

- if someone changes peer section then the interface must be restarted explicit, because nobody cares about it. For an interface section the netifd takes care of it. Can someone **please** verify this. If this is the case and you didn't want the md5sum solution please share your solution The` wg syncconf` does not work because the configuration is generate by the proto handler. That's then a hen egg problem!
https://github.com/openwrt/openwrt/pull/2625/commits/3de5ffc537966e43649e77d6f73d3aee9a72017f

- if I change the port of the listen interface, I have to adjust the firewall in a second step. This is unnecessarily. Better if the firewall itself takes care of it. Therefore I have added the wireguard service with the firewall section which improves this usability.
https://github.com/openwrt/openwrt/pull/2625/commits/2456ad413e5d164ca8682ddfbc8047f0ec251c13
https://github.com/openwrt/openwrt/pull/2625/commits/cf90637cad604ff7462c81ab152bf3dbbdcbd897

lucize commented:
this is also useful on bgp links where if the interface is restarted some bug in zebra prevents the routes installing again

adschm commented:
@blocktrron @zx2c4 Would you have another look at this one and provide a verdict on whether it should be closed?

adschm commented:
Since there has not been any activity (or interest) for one entire year and the one remark about the md5 has not been "resolved" either, I'm closing this now. Sorry, but I don't think keeping it open will be any help either.

rmilecki commented:
WireGuard config not being reloaded on peers changes is a real issue for users, e.g.:
- https://forum.openwrt.org/t/wiregurad-does-not-reload-wireguard-iface-section-on-change/47137
- https://forum.openwrt.org/t/restart-wireguard-via-cli/51935
- https://forum.openwrt.org/t/restart-wireguard-interface/58037

So we really need to fix it with this solution or another.

aparcar commented:
Another approach I suggested a long time ago: https://github.com/openwrt/openwrt/pull/4696

systemcrash commented:
Is this going to get merged this century? Or is another PR relevant?

systemcrash commented:
md5sum is entirely acceptable to detect config changes, as other procd init files do. If md5sum is a problem, then go ahead and use `sha256sum` which has better avalanche effect. 

Problem solved. Merge. Done. Everyone happy. Openwrt won. Project ended. Earth can shut down. The universe can end.

feckert commented:
Close in favour of https://github.com/openwrt/openwrt/pull/21784
